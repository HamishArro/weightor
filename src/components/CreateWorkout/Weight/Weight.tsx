import React, {useState} from 'react';
import {View, Text, TextInput, Button} from 'react-native';
import {SetProps, WeightProps} from '../types';
import {styles} from '../../../utils/styleSheet';
import {Set as SetType} from '../../../hooks/useWorkout/types';

function SetView({reps, weight, rest}: SetType) {
  return (
    <View style={styles.setContainer} testID="set-view-container">
      <Text children={reps} style={styles.text} />
      <Text children={weight} style={styles.text} />
      <Text children={rest} style={styles.text} />
    </View>
  );
}

function Set({handleAdd}: SetProps) {
  const [reps, setReps] = useState<string>('');
  const [weight, setWeight] = useState<string>('');
  const [rest, setRest] = useState<string>('');

  const handleSubmit = () => {
    setReps('');
    setRest('');
    setWeight('');
    handleAdd({
      reps: parseInt(reps, 10),
      weight: parseInt(weight, 10),
      rest: parseInt(rest, 10),
    });
  };

  return (
    <View style={styles.setContainer} testID="set-container">
      <TextInput
        testID="reps-input"
        value={reps}
        style={styles.textInput}
        onChangeText={setReps}
        placeholder="please enter number off reps"
        keyboardType="numeric"
      />
      <TextInput
        testID="weight-input"
        value={weight}
        style={styles.textInput}
        onChangeText={setWeight}
        placeholder="please enter weight"
        keyboardType="numeric"
      />
      <TextInput
        testID="rest-input"
        value={rest}
        style={styles.textInput}
        onChangeText={setRest}
        placeholder="please enter rest time"
        keyboardType="numeric"
      />
      <Button testID="save-button" title="save" onPress={handleSubmit} />
    </View>
  );
}

function Weight({handleAdd, handleBack}: WeightProps) {
  const [name, setName] = useState<string>('');
  const [musclesUsed, setMusclesUsed] = useState<string>('');
  const [workoutEffort, setWorkoutEffort] = useState<string>('');
  const [sets, setSets] = useState<SetType[]>([]);

  const handleSubmit = () => {
    handleAdd({
      name,
      musclesUsed: musclesUsed.split(','),
      workoutEffort: parseInt(workoutEffort, 10),
      sets,
    });
  };

  return (
    <View style={styles.stage} testID="weight-container">
      <Text children={'Add exercise'} style={styles.title} />
      <Button testID="back-button" title="back" onPress={handleBack} />
      <TextInput
        testID="name-input"
        value={name}
        style={styles.textInput}
        onChangeText={setName}
        placeholder="please enter name here"
      />
      <TextInput
        testID="muscles-used-input"
        value={musclesUsed}
        style={styles.textInput}
        onChangeText={setMusclesUsed}
        placeholder="please enter muscles used"
      />
      <TextInput
        testID="workout-effort-input"
        value={workoutEffort}
        style={styles.textInput}
        onChangeText={setWorkoutEffort}
        placeholder="please enter workout effort"
        keyboardType="numeric"
      />
      {sets.map(({reps, weight, rest}, index) => (
        <SetView
          key={`${reps}-${weight}-${rest}-${index}`}
          reps={reps}
          weight={weight}
          rest={rest}
        />
      ))}
      <Set handleAdd={set => setSets([...sets, set])} />
      <Button testID="add-button" title="add" onPress={handleSubmit} />
    </View>
  );
}

export default Weight;
