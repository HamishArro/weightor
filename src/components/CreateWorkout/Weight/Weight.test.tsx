import * as React from 'react';
import {describe, it, expect, beforeEach, jest} from '@jest/globals';
import {fireEvent, render, screen, act} from '@testing-library/react-native';
import Weight from './Weight';

const props = {
  handleAdd: jest.fn(),
  handleBack: jest.fn(),
};

const weightExercise = {
  name: 'weight',
  musclesUsed: ['arms'],
  workoutEffort: 50,
  sets: [{reps: 12, weight: 20, rest: 60}],
};

describe('add tests', () => {
  beforeEach(() => {
    render(<Weight {...props} />);
  });

  it('will call handle back when clicked', () => {
    act(() => fireEvent.press(screen.getByTestId('back-button')));

    expect(props.handleBack).toHaveBeenCalled();
  });

  it('can submit a form correctly', () => {
    act(() => {
      fireEvent.changeText(
        screen.getByTestId('name-input'),
        weightExercise.name,
      );
      fireEvent.changeText(
        screen.getByTestId('muscles-used-input'),
        weightExercise.musclesUsed.join(','),
      );
      fireEvent.changeText(
        screen.getByTestId('workout-effort-input'),
        JSON.stringify(weightExercise.workoutEffort),
      );
      weightExercise.sets.forEach(({reps, weight, rest}) => {
        fireEvent.changeText(
          screen.getByTestId('reps-input'),
          JSON.stringify(reps),
        );
        fireEvent.changeText(
          screen.getByTestId('weight-input'),
          JSON.stringify(weight),
        );
        fireEvent.changeText(
          screen.getByTestId('rest-input'),
          JSON.stringify(rest),
        );
      });
    });

    act(() => fireEvent.press(screen.getByTestId('save-button')));

    act(() => fireEvent.press(screen.getByTestId('add-button')));

    expect(props.handleAdd).toHaveBeenCalledWith(weightExercise);
  });

  it('matches snapshot', () => {
    expect(screen.toJSON()).toMatchSnapshot();
  });
});
