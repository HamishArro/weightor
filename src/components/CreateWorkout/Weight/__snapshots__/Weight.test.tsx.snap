// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`add tests matches snapshot 1`] = `
<View
  style={
    {
      "backgroundColor": "#dbdbdb",
      "width": "100%",
    }
  }
  testID="weight-container"
>
  <Text
    style={
      {
        "fontSize": 24,
        "paddingBottom": 18,
      }
    }
  >
    Add exercise
  </Text>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="back-button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        back
      </Text>
    </View>
  </View>
  <TextInput
    onChangeText={[Function]}
    placeholder="please enter name here"
    style={
      {
        "paddingBottom": 12,
      }
    }
    testID="name-input"
    value=""
  />
  <TextInput
    onChangeText={[Function]}
    placeholder="please enter muscles used"
    style={
      {
        "paddingBottom": 12,
      }
    }
    testID="muscles-used-input"
    value=""
  />
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    placeholder="please enter workout effort"
    style={
      {
        "paddingBottom": 12,
      }
    }
    testID="workout-effort-input"
    value=""
  />
  <View
    style={
      {
        "backgroundColor": "#F3F7F0",
      }
    }
    testID="set-container"
  >
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="please enter number off reps"
      style={
        {
          "paddingBottom": 12,
        }
      }
      testID="reps-input"
      value=""
    />
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="please enter weight"
      style={
        {
          "paddingBottom": 12,
        }
      }
      testID="weight-input"
      value=""
    />
    <TextInput
      keyboardType="numeric"
      onChangeText={[Function]}
      placeholder="please enter rest time"
      style={
        {
          "paddingBottom": 12,
        }
      }
      testID="rest-input"
      value=""
    />
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
      testID="save-button"
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          save
        </Text>
      </View>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="add-button"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        add
      </Text>
    </View>
  </View>
</View>
`;
